### ISSUE TITLE
daily file logger fails to log anything in new log file afte hour 00:00

### ISSUE THREAD

**Issue by rjvrdhn** [gabime/spdlog#1895]
As soon as the below code starts running creates log file and starts filling the log file every 1000 ms. However once the system clock reaches 00:00 it creates new logfile with new date, but fails to write anything to logfile. I even tried to flush after log entry to make sure if flush is not retained after creating new logfile. 

- OS: Ubuntu 20.04
- GCC: 9.3.0
- compiled with C++14 flag

```
#include <iostream>
#include <stdlib.h>
#include <spdlog/sinks/daily_file_sink.h>
#include <spdlog/sinks/rotating_file_sink.h>
#include <time.h>
#include <chrono>
#include <sstream>
#include <iomanip>
#include <future>
int main(int, char**) {
    std::cout << "Hello, world!\n";
    /* initialize random seed: */
    srand (time(NULL));

    auto logger = spdlog::daily_logger_mt("simpletest", "simpletest.txt", 0, 0, false, 7);
    logger->set_level(spdlog::level::debug);
    logger->flush_on(spdlog::level::debug);
    while(1)
    {
        logger->info("Log something; its a simple test");
        logger->flush();
        std::this_thread::sleep_for(std::chrono::milliseconds(1000));
    }

    return 0;
}
```
steps to reproduce:

1. set system clock to 23:57; so we do not have to wait for day change
2. run above code to see log file generated with system current date.
3. As soon as clocks runs over 00:00 and day change happens, new log file with correct date gets created, but file will be empty.

**Comment by tt4g** [comment]
Are there any errors reported to the spdlog error handler?

**Comment by rjvrdhn** [comment]
Nothing so far, I have added cout after calling logger->info. After new logfile being created, I do not see any out put as if logger->info got blocked. Here is the code I tried
```
int main(int, char**) {
    std::cout << "Hello, world!\n";
    /* initialize random seed: */
    srand (time(NULL));


    auto logger = spdlog::daily_logger_mt("simpletest", "simpletest.txt", 0, 0, false, 7);
    logger->set_error_handler([](const std::string& msg) {
        std::cerr << "my err handler: " << msg << std::endl;
        std::cout << "Error: "<<msg<<std::endl;
        throw std::runtime_error(msg);
    });

    
    logger->set_level(spdlog::level::debug);
    logger->flush_on(spdlog::level::debug);
    int k = 0;
    while(1)
    {
        try {
            logger->info("Log something; its a simple test");
            logger->flush();
            std::cout<<"Logged: "<<GetTimeStampUTCStr()<<std::endl; //After new log file created nothing prints
        }
        catch(const spdlog::spdlog_ex& ex)
        {
            std::cout << "Failed to log: " << ex.what() << std::endl;
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(60000));
    }

    return 0;
}
```

**Comment by tt4g** [comment]
Does it mean that STDOUT and STDERR are being output, but it freezes when logging is done?
If so, can you investigate which lines freeze in the debugger?

**Comment by rjvrdhn** [comment]
I did little bit of debugging to see where it gets blocked. Of course the code I am in was in **daily_file_sink.h**. Its inside `delete_old() `function, there is function call `filename()` to get current file name . In filename function it tries to acquire lock with `std::lock_guard<Mutex> lock(base_sink<Mutex>::mutex_);`. This lock call getting blocked. I could not find place where its already locked so far.

**Comment by tt4g** [comment]
It may be a known issue mentioned in this comment: https://github.com/gabime/spdlog/issues/1879#issuecomment-800618950

If `std::thread::hardware_concurrency()` returns `1`, your machine does not support multi-thread.

**Comment by tt4g** [comment]
Oh, I remember now.
The deadlock of calling `filename()` in `delete_old_()` was fixed (#1571) over a year ago.

This fix is included in version v1.6.1 and later.

**Comment by rjvrdhn** [comment]
you are right, I had the version 1.6.0. Updated to latest and it worked. Thanks
